Check existing storage quotas and adjust UI appropriately

zherman0 opened the pull request:
Use the storage quota features from the ResourceQuotas to let users know when they have exceeded their allotment of claims or when they exceed their max storage size.

zherman0 commented:
How attach pvc page looks when claims are exhausted:
![quota_attach-nocreate](https://cloud.githubusercontent.com/assets/18728857/22574075/cf73c94e-e969-11e6-8b0d-33ff4ac67ee2.png)
How storage page looks when claims are exhausted:
![quotas_storage-nocreate](https://cloud.githubusercontent.com/assets/18728857/22574096/f665302e-e969-11e6-8bc8-533bbc9ec390.png)
Create storage page when claims exhausted:
![quota-claimsusedup](https://cloud.githubusercontent.com/assets/18728857/22574102/0744b41e-e96a-11e6-8d14-a92639f7d35f.png)
Create storage when max storage size exceeded:
![quota-maxstorageexceeded](https://cloud.githubusercontent.com/assets/18728857/22574113/1bf236b6-e96a-11e6-9642-14d21fe19a82.png)

spadgett commented:
@jwforres Might be good if you look at this since you added other quota warnings

jwforres commented on app/scripts/controllers/attachPVC.js:
```diff
@@ -133,6 +135,16 @@ angular.module('openshiftConsole')
             }
           });
+
+          DataService.list('resourcequotas', { namespace: $scope.projectName }, function(quotaData) {
```
any time you check resourcequotas, you also have to check clusterresourcequotas

jwforres commented on app/scripts/filters/resources.js:
```diff
@@ -1137,6 +1138,7 @@ angular.module('openshiftConsole')
         'openshift.io/imagestreamsize': 'image stream size',
         'openshift.io/projectimagessize': 'project image size',
         'persistentvolumeclaims': 'persistent volume claims',
+        'requests.storage': 'Storage (Request)',
```
this one should have a lower case 'r' on request

jwforres commented on app/scripts/directives/oscPersistentVolumeClaim.js:
```diff
@@ -134,6 +155,17 @@ angular.module("openshiftConsole")
+          }
+
+          scope.storageQuota = QuotaService.getStorageQuotas(quotas);
+          console.log("Storage Quotas",scope.storageQuota);
```
if you are wanting to keep this debug logging it needs to use the Logger service instead, we dont check in console.log stmts

jwforres commented on app/scripts/directives/oscPersistentVolumeClaim.js:
```diff
@@ -22,6 +23,10 @@ angular.module("openshiftConsole")
+        scope.maxNumberClaims = 0;
+        scope.maxStorage = 0;
+        scope.claimCountUsed = 0;
+        scope.claimStorageused = 0;
```
i dont see where this variable directly on the scope is getting used, all the refs I see are to either a local var or one that is on scope.storageQuota

jwforres commented on app/scripts/services/quota.js:
```diff
@@ -277,6 +282,45 @@ angular.module("openshiftConsole")
+    // and the amount of claims and storage already used
+
+    var getStorageQuotas = function(quotas) {
+      // Currently, there can be multiple quotas per namespace (bug?)
```
nope, intentional. as many quotas as you want, plus you have to deal with clusterresourcequotas.

jwforres commented on app/scripts/services/quota.js:
```diff
@@ -277,6 +282,45 @@ angular.module("openshiftConsole")
+    var getStorageQuotas = function(quotas) {
+      // Currently, there can be multiple quotas per namespace (bug?)
+      // So let's go through them and find the most restrictive
+      var maxNumberClaims = 0;
```
the logic in here doesn't seem right to me, how is it handling if none of the quotas have storage restrictions, shouldn't these values be null and not 0 in that case.

spadgett commented on app/scripts/services/quota.js:
```diff
@@ -277,6 +282,45 @@ angular.module("openshiftConsole")
+      var maxNumberClaims = 0;
+      var maxStorage = 0;
+      var claimCountUsed = 0;
+      var claimStorageUsed = 0;
```
I suggest one

```js
var storageQuotas = {};
```

variable, then these values directly in `storageQuotas`, which you can then just return at the end of the function.

spadgett replied:
Defaults should probably be undefined instead of 0 anyway so you can tell the difference between a zero value and unset.

spadgett commented on app/views/attach-pvc.html:
```diff
@@ -74,7 +74,8 @@ <h2 class="text-center">No persistent volume claims.</h2>
                       <div ng-if="project && ('persistentvolumeclaims' | canI : 'create')" class="help-block">
-                        Select storage to use or <a ng-href="project/{{project.metadata.name}}/create-pvc" >create storage</a>.
+                        Select storage to use<span ng-if="!outOfClaims"> or <a ng-href="project/{{project.metadata.name}}/create-pvc" >create storage</a>.</span>
+                        <span ng-if="outOfClaims">. Due to exceeded quotas, new storage cannot be create.</span>
```
Maybe just "You cannot create new storage since you are at quota."

Although @jwforres is a better wordsmith than me.

spadgett commented on app/views/storage.html:
```diff
@@ -30,11 +30,17 @@
-                  <a ng-href="project/{{project.metadata.name}}/create-pvc" class="btn btn-default">Create Storage</a>
+                  <a ng-if="!outOfClaims" ng-href="project/{{project.metadata.name}}/create-pvc" class="btn btn-default">Create Storage</a>
+                  <div ng-if="outOfClaims" class="has-error">
+                    <span class="help-block">Exceeded quotas prevent creating new storage</span>
```
So I don't think it should be an error here. Being at quota doesn't mean there's a problem. We have an alert we show when at quota on the overview page you might look at with a "Don't show me again" link.

zherman0 commented:
@jwforres @spadgett @erinboyd - I have refactored this code to include cluster quotas, reuse existing code in the quota library, and make the over all logic more robust.  Let me know what else you see. Thanks for all the input so far; it's really good stuff.

zherman0 commented:
@spadgett @jwforres - I know you guys are super busy this week, but I just want to see if we can get this merged by Friday's deadline. Thanks.

jwforres commented:
This is a significant set of changes and the equivalent work for the other quota checks was a Trello story that had to go through QE, I don't think this can go in as a bug fix this late in the release.  We need a card on the UI board for 3.6 for QE to test.

erinboyd commented:
@jwforres Fair enough. Can we just create the card on our storage board, or do you want it tracked via the UI board?

erinboyd commented:
@jwforres I don't disagree it's late in the game. But can we see if the QE team has the cycles to test it since it's only a few things on the storage side and they will have already tested the quota stuff you added?

jwforres commented:
Works for me

On Wed, Feb 8, 2017 at 2:12 PM, Erin A Boyd <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> Fair enough. Can we just create
> the card on our storage board, or do you want it tracked via the UI board?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/1217#issuecomment-278430767>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7fxHnoua3TortGoDtPySOQ3i1OW8ks5rahOzgaJpZM4L1zYO>
> .
>

zherman0 commented:
@jwforres - Jessica can you add me (zacherman1) to your trello board so I can create a card. Thanks.

pweil- commented:
At this point we're working things on the blocker list and stabilizing, anything that isn't considered part of that should wait until 3.5.1 (if it cannot wait until 3.6).  Let's get the card created though, QE will need a set of test cases.

jwforres commented on app/scripts/services/quota.js:
```diff
@@ -277,6 +280,50 @@ angular.module("openshiftConsole")
+          }
+
+          //Get the amount used so far
+          var used = usedStorage(quota);
```
this logic isn't going to work if you've got multiple cluster quotas, imagine you've got projects A, B, and C.  clusterquota-1 covers projects A and B, clusterquota-2 covers projects B and C.   Consume some storage resources in project A, consume a different amount of storage resources in C.  If you are then in project B you will see two clusterquotas with different amounts of "used" quota.

jwforres replied:
actually the maxStorage part is problematic for similar reasons

jwforres commented on app/views/directives/osc-persistent-volume-claim.html:
```diff
@@ -186,6 +193,11 @@
            </div>
+           <div ng-if="persistentVolumeClaimForm.capacity.$error.underStorage" class="has-error">
+             <span class="help-block">
+               Storage quota will be exceeded - Max allocated: {{storageQuota.maxStorage}} Currently used: {{storageQuota.claimStorageUsed}}.
```
this unfortunately can't be this clear cut when throwing cluster quotas into the mix, I think you will need to just link them to the quota page instead of trying to come up with these numbers

erinboyd replied:
@zherman0 what is the status of the new logic? I am not seeing a commit for this

erinboyd commented:
@jwforres does this look accurate: https://docs.google.com/a/redhat.com/document/d/1Y9P6Ffa5cAF9MqlBtLvSXlmiRzKDxTfuuoYFIeWnF7w/edit?usp=sharing

jwforres commented:
@erinboyd @zherman0 is the still intended to go into 3.6?

zherman0 commented:
Yes, that is the plan. I will get that branch rebased with the other few
changes by EOD.
Thanks.

On Mon, Apr 17, 2017 at 7:23 AM, Jessica Forrester <notifications@github.com
> wrote:

> @erinboyd <https://github.com/erinboyd> @zherman0
> <https://github.com/zherman0> is the still intended to go into 3.6?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/1217#issuecomment-294484801>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR3HmRXshDNtSE4WeTT5UiGlY97xEQIRks5rw2frgaJpZM4L1zYO>
> .
>



-- 

Zac Herman

Senior Software Engineer

Red Hat

<https://www.redhat.com>

zherman@redhat.com
<https://red.ht/sig>

zherman0 commented:
@jwforres @spadgett - Heys guys, I rebased and refactored this to handle quotas and cluster quotas.  Please take and a look and let's try to get the merge asap. Thanks.

jwforres commented on app/scripts/services/quota.js:
```diff
@@ -277,6 +279,45 @@ angular.module("openshiftConsole")
 
+    // Same as above but only looking at storage items: requests.storage, persistentvolumeclaims
+    //   Warn if you are at quota or over any storage quota for any resource.
+    var isAnyStorageQuotaExceeded = function(quotas, clusterQuotas, requestedStorage) {
```
instead of copying the function above I would rather have an optional third parameter to isAnyQuotaExceeded that is a list of quota keys to check, then your if check for whether you check that particular quotaKey is something like 

```
if (!typesToCheck || _.includes(typesToCheck, quotaKey) {
  hard = usageValue(hard);
  // etc...
```

then if you wanted to you could add a function for "isAnyStorageQuotaExceeded" that looks more like:
```
var isAnyStorageQuotaExceeded = function(quotas, clusterQuotas) {
  return isAnyQuotaExceeded(quotas, clusterQuotas, ['requests.storage', 'persistentvolumeclaims]');
};
```

The only case where you were actually using requestedStorage was for the field validity check on requests.storage.  For that purpose I would rather see a generic function written that checks a particular quotaKey's `hard <= (used + request)`

jwforres commented on app/views/directives/osc-persistent-volume-claim.html:
```diff
@@ -1,5 +1,12 @@
+    <div ng-if="persistentVolumeClaimForm.capacity.$error.outOfStorage" class="has-error">
+      <div class="alert alert-danger">
+        <span class="pficon pficon-error-circle-o" aria-hidden="true"></span>
+        <strong>Quota limit has been reached. New storage cannot be created. To increase your quota, contact your administrator.</strong>
```
same comment as above, make this "Storage quota limit has been reached. You will not be able to create any new storage."

jwforres commented on app/views/directives/osc-persistent-volume-claim.html:
```diff
@@ -180,6 +187,11 @@
                Can't be greater than {{limits.max | usageWithUnits : 'storage'}}.
              </span>
            </div>
+           <div ng-if="persistentVolumeClaimForm.capacity.$error.outOfStorage" class="has-error">
```
given the context this validity is used in, it is strange that my PVC quota being exceeded would flag an error on this field.  i would expect you to only be checking requests.storage here

zherman0 replied:
I created 2 flags here, one for out of claims and one for out of storage

jwforres replied:
the variable names are a little confusing given what they are actually doing... "outOfClaims" is really am I out of either claims or storage, and "outOfStorage" is really will my request exceed my storage quota.  I think at least changing "outOfStorage" to "willExceedStorage" would help

jwforres commented on app/views/storage.html:
```diff
@@ -30,11 +30,11 @@
               <div class="section-header page-header-bleed-right page-header-bleed-left">
                 <div class="hidden-xs pull-right" ng-if="project && ('persistentvolumeclaims' | canI : 'create')">
-                  <a ng-href="project/{{project.metadata.name}}/create-pvc" class="btn btn-default">Create Storage</a>
+                  <a ng-if="!outOfClaims" ng-href="project/{{project.metadata.name}}/create-pvc" class="btn btn-default">Create Storage</a>
```
this situation might justify a disabled button over a missing action. thoughts @spadgett ?  If we hide the buttons then we need to take that into account in the if check when considering whether the dropdown button (kebab at mobile) should be visible.  Meaning if this individual action was hidden because of the outOfClaims check, resulting in an empty dropdown, we would want to make sure the dropdown button is hidden.

zherman0 replied:
@spadgett @jwforres - I can change this to disabled instead, just let me know.

zherman0 commented:
@jwforres - I made your recommended changes. Still waiting to hear what to do about the button.

jwforres commented:
lets go with disabled

On Thu, Apr 27, 2017 at 12:23 PM, Zac Herman <notifications@github.com>
wrote:

> *@zherman0* commented on this pull request.
> ------------------------------
>
> In app/views/storage.html
> <https://github.com/openshift/origin-web-console/pull/1217#discussion_r113741904>
> :
>
> > @@ -30,11 +30,11 @@
>              <div class="col-md-12">
>                <div class="section-header page-header-bleed-right page-header-bleed-left">
>                  <div class="hidden-xs pull-right" ng-if="project && ('persistentvolumeclaims' | canI : 'create')">
> -                  <a ng-href="project/{{project.metadata.name}}/create-pvc" class="btn btn-default">Create Storage</a>
> +                  <a ng-if="!outOfClaims" ng-href="project/{{project.metadata.name}}/create-pvc" class="btn btn-default">Create Storage</a>
>
> @spadgett <https://github.com/spadgett> @jwforres
> <https://github.com/jwforres> - I can change this to disabled instead,
> just let me know.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/1217#discussion_r113741904>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7QyYOKrPyVpdyabfov3sXIdXwqEDks5r0METgaJpZM4L1zYO>
> .
>

zherman0 commented:
@spadgett @jwforres - Switched the button to disabled instead of hiding it

jwforres commented on app/scripts/services/quota.js:
```diff
@@ -260,18 +262,55 @@ angular.module("openshiftConsole")
+    };
+
+   // Check if requested storage for a new claim will exceed any quotas if attempted
+    var willRequestStorageQuotaBeExceeded = function(quotas, clusterQuotas, requestedStorage) {
```
this doesnt need to be specific to storage quota, i was suggesting this be "willRequestExceedQuota", take four parameters of quotas, clusterQuotas, quotaKey, request and then we could reuse it for other stuff later.  You should really make this take request in the format that it is sent to the API which is value+unit, so you'll want to run usageValue on it.

jwforres commented on app/scripts/directives/oscPersistentVolumeClaim.js:
```diff
@@ -85,6 +86,14 @@ angular.module("openshiftConsole")
+        var validateQuota = function() {
+          var requestedStorage = scope.claim.amount && usageValue(scope.claim.amount + scope.claim.unit);
+          var outOfClaims = QuotaService.isAnyStorageQuotaExceeded(scope.quotas, scope.clusterQuotas);
+          var outOfStorage = QuotaService.willRequestStorageQuotaBeExceeded(scope.quotas, scope.clusterQuotas, requestedStorage);
```
so pass scope.claim.amount + scope.claim.unit here instead of the converted value, and let the quota service method handle the usageValue conversion stuff

zherman0 commented:
@jwforres - I made a generic function now to check if a quota will be exceeded.

jwforres commented on app/scripts/services/quota.js:
```diff
@@ -260,23 +262,59 @@ angular.module("openshiftConsole")
+        if (!requestedQuotaKey) {
+          return false;
+        }
+        return _.some(q.hard, function(hard, quotaKey) {
```
so you know the exact key you want to check, instead of iterating over every quotaKey with _.some can't you just directly access it?

```
var hard = _.get(q.hard, quotaKey);
hard = usageValue(hard);
// etc...
```

zherman0 commented:
@jwforres - I tweaked the generic willRequestExceedQuota function so it only looks at the requested quota key.  Thanks for the great feedback.

jwforres commented on app/scripts/services/quota.js:
```diff
@@ -260,23 +262,53 @@ angular.module("openshiftConsole")
+        }
+        var used = usageValue(_.get(q, ['used', requestedQuotaKey]));
+        if (!used) {
+          return false;
```
sorry one more thing :)
even if you don't have any used, your requested value might still exceed your quota, so this check need to be
```
if (!used) {
  return hard <= value;
}
```

zherman0 replied:
@jwforres - Wow, great catch, I did not have a testcase for that.  See how it looks now.

jwforres commented:
LGTM now, we can merge on monday when master is reopened for features

zherman0 commented:
Thanks.

On Thu, May 11, 2017 at 3:03 PM, Jessica Forrester <notifications@github.com
> wrote:

> LGTM now, we can merge on monday when master is reopened for features
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/1217#issuecomment-300916349>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR3Hmd8OkTce_2Wc9QFZ11I-JPwSl6q0ks5r43ewgaJpZM4L1zYO>
> .
>



-- 

Zac Herman

Senior Software Engineer

Red Hat

<https://www.redhat.com>

zherman@redhat.com
<https://red.ht/sig>

jwforres commented:
[merge]

jwforres commented:
@zherman0 getting a diff on vendor.js, you'll need to rebase this on master and clean-deps, install-deps and re-build

spadgett commented:
@zherman0 Failed on a diff in vendor.js. Can you try `bower install && grunt build` and update your commit?

zherman0 commented:
@jwforres - I rebased with master and updated the code, hopefully that resolves things

jwforres commented:
[merge]
